feat: Add password & PUT endpoint to Users

zxkevin opened the pull request:
Added new migrations to update Users table with password column
Seed file added to update pre-existing seeds with a password.
POST endpoint for Users now requires a password.
PUT endpoint for {userID} added for update users functionality.
Bcrypt is used to salt and hash password for secure password storage.

#87 #51

zxkevin committed:
feat: Add password & PUT endpoint to Users

Added new migrations to update Users table with password column
Seed file added to update pre-existing seeds with a password.
POST endpoint for Users now requires a password.
PUT endpoint for {userID} added for update users functionality.
Bcrypt is used to salt and hash password for secure password storage.

#87 #51

zxkevin commented on packages/hackmcx-api/src/db/seeds/2022-03-29_1757_update_demo_users.cjs:
```diff
@@ -0,0 +1,5 @@
+function seed(knex) {
+    return knex('users').update('password', '$2a$12$A0HScvmVMgyRKcy/BYm/tuYi7IetjaYNpzU/gVHBmtHu3EtWUPshq');
```
"password123"

3dxp commented on packages/hackmcx-api/src/controllers/users.controller.js:
```diff
@@ -22,38 +23,49 @@ export async function getUsersByUserId(req, res){
 
-export async function postUser(req, res){
+function userValidation(req){
     const emptyUsername = !req.body.username || /^\s*$/.test(req.body.username);
```
I think at this point, checking for null or whitespace should be a helper function, with the regex compiled.

3dxp commented on packages/hackmcx-api/src/controllers/users.controller.js:
```diff
@@ -22,38 +23,49 @@ export async function getUsersByUserId(req, res){
     const notPicture = !req.body.imageUrl || !/(https?:\/\/.*\.(?:png|jpg|gif|svg))/i.test(req.body.imageUrl);
+    const noGivenImageUrl = !req.body.imageUrl;
 
     if(emptyUsername){
```
I think these validation checks can be factored out, and maybe applied elsewhere. 

Eventually we might want to look at just adding some reasonable exception handling and breaking all this validation out of the controller.

3dxp commented on packages/hackmcx-api/src/controllers/users.controller.js:
```diff
@@ -22,38 +23,49 @@ export async function getUsersByUserId(req, res){
+        return;
     }
 
+    const hashPass = await bcrypt.hash(req.body.password, 12);
```
Why is this a `const`?

3dxp commented on packages/hackmcx-api/package.json:
```diff
@@ -9,11 +9,12 @@
   },
   "dependencies": {
+    "bcrypt": "^5.0.1",
+    "cors": "^2.8.5",
```
It seems like the only thing changing here is the addition of bcrypt. We might want to consider removing the other changes, even it they do reorganize the imports in alphabetical order.

acampb65 reviewed (changes requested):
I think preexisting seeds need to be modified. The update demo users seed works fine if rows already exist, but when running seed data on empty database MySQL gets mad:
```
code: 'ER_NO_DEFAULT_FOR_FIELD',
    errno: 1364,
    sqlState: 'HY000',
    sqlMessage: "Field 'password' doesn't have a default value",
    sql: "insert into `users` (`first_name`, `imageUrl`, `last_name`, `username`) values ('first_name1', 'https://upload.wikimedia.org/wikipedia/commons/thumb/2/20/MJd1-.svg/100px-MJd1-.svg.png', 'last_name1', 'username1'), ('first_name2', 'https://upload.wikimedia.org/wikipedia/commons/thumb/1/17/MJd2v1-.svg/100px-MJd2v1-.svg.png', 'last_name2', 'username2'), ('first_name3', 'https://upload.wikimedia.org/wikipedia/commons/thumb/0/04/MJd3e-.svg/100px-MJd3e-.svg.png', 'last_name3', 'username3')"
```

acampb65 commented:
PUT user responds with 201 even if username doesn't exist.

example:
```
curl -X PUT -i http://localhost:3003/api/v1/users/nonExistentUsername \
-H 'Content-Type: application/json' \
-d '{"username":"user","password":"pass","firstname":"first","lastname":"last"}'
       
HTTP/1.1 201 Created
X-Powered-By: Express
Access-Control-Allow-Origin: *
Location: /api/v1/users/undefined
Date: Wed, 30 Mar 2022 18:11:43 GMT
Connection: keep-alive
Keep-Alive: timeout=5
Content-Length: 0

```

zxkevin committed:
fix: Update old seed, PUT inserts and updates

Updated the old seed data so that a password field exists
so that seeding will not break
Updated the PUT endpoint so that it will create new users if not
already in database and update users if in database

#87 #51

acampb65 reviewed (approved):
LGTM
